src/pudl/analysis/demand_anomalies.py

```diff
@@ -0,0 +1,1316 @@
+        # Only flag unflagged values
+        mask = mask & ~np.isnan(self.x)
+        self.flags[mask] = flag
+        self.flagged.append(flag)
+        # Null flagged values
+        self.x[mask] = np.nan
+        # Clear cached metrics
+        for name in dir(self):
+            attr = getattr(self, name)
+            if hasattr(attr, 'cache_clear'):
+                attr.cache_clear()
+
+    def unflag(self, flags: Iterable[str]) -> None:
+        """
+        Unflag values.
+
+        Unflags values by restoring their original values and removing their flag.
+
+        Args:
+            flags: Flag names.
+        """
+        mask = np.isin(self.flags, flags)
+        self.flags[mask] = None
+        self.x[mask] = self.xi[mask]
+        self.flagged = [f for f in self.flagged if f not in flags]
+
+    def flag_negative_or_zero(self) -> None:
+        """Flag negative or zero values (NEGATIVE_OR_ZERO)."""
+        mask = self.x <= 0
+        self.flag(mask, "NEGATIVE_OR_ZERO")
+
+    def flag_identical_run(self, length: int = 3) -> None:
+        """
+        Flag the last values in identical runs (IDENTICAL_RUN).
+
+        Args:
+            length: Run length to flag.
+                If `3`, the third (and subsequent) identical values are flagged.
+
+        Raises:
+            ValueError: Run length must be 2 or greater.
+        """
+        if length < 2:
+            raise ValueError("Run length must be 2 or greater")
+        mask = np.ones(self.x.shape, dtype=bool)
+        mask[0] = False
+        for n in range(1, length):
+            mask[n:] &= self.x[n:] == self.x[:-n]
+        self.flag(mask, "IDENTICAL_RUN")
+
+    def flag_global_outlier(self, medians: Union[int, float] = 9) -> None:
+        """
+        Flag values greater or less than n times the global median (GLOBAL_OUTLIER).
+
+        Args:
+            medians: Number of times the median the value must exceed the median.
+        """
+        median = np.nanmedian(self.x, axis=0)
+        mask = np.abs(self.x - median) > np.abs(median * medians)
+        self.flag(mask, "GLOBAL_OUTLIER")
+
+    def flag_global_outlier_neighbor(self, neighbors: int = 1) -> None:
+        """
+        Flag values neighboring global outliers (GLOBAL_OUTLIER_NEIGHBOR).
+
+        Args:
+            neighbors: Number of neighbors to flag on either side of each outlier.
+
+        Raises:
+            ValueError: Global outliers must be flagged first.
+        """
+        if "GLOBAL_OUTLIER" not in self.flagged:
+            raise ValueError("Global outliers must be flagged first")
+        mask = np.zeros(self.x.shape, dtype=bool)
+        outliers = self.flags == "GLOBAL_OUTLIER"
+        for shift in range(1, neighbors + 1):
+            # Neighbors before
+            mask[:-shift][outliers[shift:]] = True
+            # Neighors after
+            mask[shift:][outliers[:-shift]] = True
+        self.flag(mask, "GLOBAL_OUTLIER_NEIGHBOR")
+
+    @functools.lru_cache(maxsize=2)
```

Comment:
Docs: https://docs.python.org/3/library/functools.html#functools.lru_cache

The "metric" methods (e.g. `rolling_median`, etc) are used by Tyler Ruggles' more complex anomaly detection routines. In some cases, a `flag_*` method needs access to both a metric A and another metric B which depends on metric A. To make this more transparent, I broke out each metric into its own documented method. However, to avoid recomputing the same metric from the same `flag_*` method, I use a cache on the reused metric method, which is then cleared upon the completion of the `flag_*` method.